Fix the bug when using gdbserver

lebr0nli opened the pull request:
If we execute the `monitor` command in `is_debug_probe()`, the GDB with normal GDBserver will ~~stuck~~ after first `continue`.(Edit: it didn't freeze, instead, it showed an additional newline and won't show context because `pwndbg.gdblib.proc.thread_is_stopped` is False when running `gdb.prompt_hook`)
> To reproduce this bug, you can use pwntools' `gdb.debug()`

To avoid this, we check if `monitor help` showed that we are using normal GDBserver.

lebr0nli committed:
Fix the bug when using gdbserver

If we execute the `monitor` command in `is_debug_probe()`, the GDB with normal GDBserver will stuck after first `continue`.

To avoid this, we check if `monitor help` showed that we are using GDBserver.

disconnect3d commented:
Fwiw interestingly doing just:
```
gdbserver 127.0.0.1:1234 /bin/ls
```
in one terminal, and then: `gdb -ex "target remote :1234"` in another terminal does not make GDB/Pwndbg stuck/hang on this.

disconnect3d commented:
@lebr0nli I can't reproduce it with `from pwn import *; x = gdb.debug('/bin/ls'); x.interactive()` too. 

What are the steps to reproduce it? Or what is your GDB version?

lebr0nli commented:
> @lebr0nli I can't reproduce it with `from pwn import *; x = gdb.debug('/bin/ls'); x.interactive()` too.

Ahh, interesting, seems like `/bin/ls` won't be affected by this bug.

> What are the steps to reproduce it?

1. `vim test.c`:
```c
#include <stdio.h>

int main(){
    printf("abcd\n");
    return 0;
}
```
2. `gcc test.c -o test`
3. `python -c "from pwn import *; x = gdb.debug('test',gdbscript='b main');x.interactive()"`
If you are using tmux, after using `continue` in the panel of GDB, it should stuck. You will need to `Ctrl+C` to make GDB work again.

 > Or what is your GDB version?

I tested this on 11.2 and 12.1, both have same issue.

disconnect3d commented:
Hmm it works for me on GDB 10.2:
<img width="1410" alt="image" src="https://user-images.githubusercontent.com/10009354/207052400-3eb3b277-0fbb-40bc-ac1a-2b8c0bf7f0c8.png">

lebr0nli commented:
> Hmm it works for me on GDB 10.2:

Hmm, weird, here's mine after I `continue`:

<img width="1397" src="https://user-images.githubusercontent.com/61896187/207053627-ac2d631f-52ec-4979-840e-ef7f3f7dcbdb.png">

somehow I need to `Ctrl+C` to make it show the prompt and ctx

My version:
```
pwndbg> version
Gdb:      12.1
Python:   3.8.10 (default, Nov 14 2022, 12:59:47)  [GCC 9.4.0]
Pwndbg:   1.1.1 build: f49aa00
Capstone: 4.0.1024
Unicorn:  2.0.1
``` 

I'll test it on older GDB later to see what's going on.

disconnect3d commented:
Btw what distro/terminal is it? @szymex73 tested on GDB 12.1 and it works for him 🤔

szymex73 commented:
It seems fine for me 
![screensshot_2022-12-12_14-28-54](https://user-images.githubusercontent.com/5494832/207056965-5313ed4a-3bd6-4ac7-9dbd-7ec6a2d0c8da.png)
Also, in your command ensure to run `./test` and not `test` as it will run test from your PATH

disconnect3d commented:
Oh my bad, I used old version of Pwndbg accidentally. I can reproduce your issue partially, or to be specific:

<img width="1257" alt="image" src="https://user-images.githubusercontent.com/10009354/207057645-3bc3d4f5-e4d2-45a0-acc3-4f942414377a.png">

When you see the 'ls' this is my input.
Sooo it seems like a newline was printed by GDB/us somehow but we still hit/stopped properly on the brekapoint

szymex73 commented:
I've updated to the latest commit from `dev` and the freeze does indeed appear on my machine with the steps from @lebr0nli 
![screensshot_2022-12-12_14-37-18](https://user-images.githubusercontent.com/5494832/207058466-95b1a357-4003-4271-92c1-8c3094bf15cd.png)
(Notice the Quit after a Ctrl-C to get back the prompt)

szymex73 commented:
Okay so a bit more testing/investigation and on my end the hang turned out to be a newline like @disconnect3d has shown, but that seemed to unearth some other weird behavior that I'm not sure what it is yet.
It's probably a bug within gdb itself though as it persists (and is easier to reproduce, I suspect because of delays pwndbg might be introducing between the commands being executed) without pwndbg loaded. Gdb seems to crash when you run the continue command twice very quickly (gdb/gdbserver being run in the same way as above, through the pwntools gdb.debug call) but I haven't been able to get any trace because it just exits and the tmux pane disappears :upside_down_face:.

disconnect3d commented:
Yeah, it seems this halting may be a GDB or Pwntools issue? See this with no Pwndbg loaded:
```
 python -c "from pwn import *; x = gdb.debug('./test',gdbscript='b main \n c \n c ');x.interactive()"
 ```

disconnect3d commented:
This can be reproduced with:
```
gdbserver --multi --no-disable-randomization 127.0.0.1:1234 ./test
```
And:
```
gdb -q ./test -ex 'file ./test' -ex 'target remote 127.0.0.1:1234' -ex 'file ./test' -ex 'b main' -ex 'c' -ex 'c'
```

disconnect3d commented:
@lebr0nli Can you please double check/confirm if it is really halting/freezing GDB+Pwndbg for you, or is it just that we are displaying additional newline after prompt is displayed?

lebr0nli committed:
Update the comment about the bug caused by `monitor`

The GDB didn't freeze, but somehow showed an additional newline and failed to show the context because `pwndbg.gdblib.proc.thread_is_stopped` is False when running `gdb.prompt_hook`.

lebr0nli commented:
> @lebr0nli Can you please double check/confirm if it is really halting/freezing GDB+Pwndbg for you, or is it just that we are displaying additional newline after prompt is displayed?

You're right, it somehow displays an additional newline in the end and not showed the context because `pwndbg.gdblib.proc.thread_is_stopped` is `False` when running `gdb.prompt_hook` if we run a failed `monitor` command in the background.
(maybe is because the `monitor` command crashed but gdb thinks the command is still executing? IDK, just a random guess)
I guess this probably is a bug of gdb/gdbserver, but accidentally trigger by our failed `monitor` command lol.

I'll update this information in the comment.

disconnect3d commented:
@alufers Can you double check if doing the change in this PR still works for your debug bridge?

alufers commented:
Yes, it still appears to be working. Calling monitor twice does seem to be a hack, though. 

These are the outputs for various monitor commands using the Blackmagic probe and J-Link, for reference:

```sh
(gdb) monitor help
General commands:
	version -- Display firmware version info
	help -- Display help for monitor commands
	jtag_scan -- Scan JTAG chain for devices
	swdp_scan -- Scan SW-DP for devices
	auto_scan -- Automatically scan all chain types for devices
	frequency -- set minimum high and low times
	targets -- Display list of available targets
	morse -- Display morse error message
	halt_timeout -- Timeout (ms) to wait until Cortex-M is halted: (Default 2000)
	connect_rst -- Configure connect under reset: (enable|disable)
	reset -- Pulse the nRST line - disconnects target
	tdi_low_reset -- Pulse nRST with TDI set low to attempt to wake certain targets up (eg LPC82x)
	rtt -- enable|disable|status|channel 0..15|ident (str)|cblock|ram|poll maxms minms maxerr
	traceswo -- Start trace capture, Manchester mode: (decode channel ...)
	heapinfo -- Set semihosting heapinfo
(gdb) monitor version
Black Magic Probe(BlackPillV2) v1.8.0-1036-g91a5ea91, Hardware Version 0
Copyright (C) 2022 Black Magic Debug Project
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

(gdb) monitor
Black Magic Probe(BlackPillV2) v1.8.0-1036-g91a5ea91, Hardware Version 0
Copyright (C) 2022 Black Magic Debug Project
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
```


```sh
pwndbg> monitor help
SEGGER J-Link GDB Server V7.82d

Available remote commands are:

  cp15            clrbp           ci              clrwp           
  cd              device          flash device    endian          
  ExcludeFlashCacheRange  flash breakpoints  flash cpuclock  flash download  
  flash erase     getargs         go              halt            
  IsHalted        interface       if              jtagconf        
  loadbin         rtos            rtosconf        MemU16          
  WriteU16        MemU32          WriteU32        mww             
  long            MemU8           WriteU8         reg             
  regs            remoteport      port            reset           
  STRACE Config   STRACE Read     STRACE Start    STRACE Stop     
  SWO DisableTarget  SWO EnableTarget  SWO GetMaxSpeed  SWO GetSpeedInfo  
  SWOSpeed        SWO Start       SWO Stop        select          
  semihosting ARMSWI  semihosting disable  semihosting enable  semihosting IOClient  
  semihosting BreakOnError  semihosting ThumbSWI  setargs         setbp           
  bi              ReadMemAP       WriteDP         WriteAP         
  ReadAPEx        ReadDP          ReadAP          bdh             
  bd              bdb             sleep           speed           
  step            si              waithalt        wh              
  wice            rmib            WriteCSR        ReadCSR         
pwndbg> monitor
SEGGER J-Link GDB Server V7.82d

Available remote commands are:

  cp15            clrbp           ci              clrwp           
  cd              device          flash device    endian          
  ExcludeFlashCacheRange  flash breakpoints  flash cpuclock  flash download  
  flash erase     getargs         go              halt            
  IsHalted        interface       if              jtagconf        
  loadbin         rtos            rtosconf        MemU16          
  WriteU16        MemU32          WriteU32        mww             
  long            MemU8           WriteU8         reg             
  regs            remoteport      port            reset           
  STRACE Config   STRACE Read     STRACE Start    STRACE Stop     
  SWO DisableTarget  SWO EnableTarget  SWO GetMaxSpeed  SWO GetSpeedInfo  
  SWOSpeed        SWO Start       SWO Stop        select          
  semihosting ARMSWI  semihosting disable  semihosting enable  semihosting IOClient  
  semihosting BreakOnError  semihosting ThumbSWI  setargs         setbp           
  bi              ReadMemAP       WriteDP         WriteAP         
  ReadAPEx        ReadDP          ReadAP          bdh             
  bd              bdb             sleep           speed           
  step            si              waithalt        wh              
  wice            rmib            WriteCSR        ReadCSR         
pwndbg> monitor version
Target does not support this command.
```

lebr0nli committed:
Reduce unnecessary `monitor` called if possible

lebr0nli commented:
@alufers, thanks for providing the full output of the `monitor` and `monitor help` commands!
> Calling monitor twice does seem to be a hack, though.

I did a new commit to reduce unnecessary `monitor` called if possible, could you double-check if the new commit also works for you? Thanks!
(After the new commit we should only call monitor twice if we are using some remote server that support `monitor help` and the server is not SEGGER J-Link GDB Server or normal gdbserver)

alufers commented:
Seems fine by me!
